28/sqli session (#29)
Cov-ComSec/Cov-ComSec.github.io: content/posts/2021_sqli.md

---
title: "SQL Injection"
date: 2022-03-02T12:40:28+01:00
draft: false
author: "Jack Orcherton"

> authorLink: "https://jackorcherton.github.io"

Reviewer: Oh cool. Never knew this font matter existed

Author: yep think I've done it to most of mine, feel free to add to yours.

The theme the site uses is one based off of Hugo Love It (however I have made a few changes here and there) -> the full front matter options can be seen at https://hugoloveit.com/theme-documentation-content/#front-matter

description: "An introductory session into SQL Injection"
tags: ["SQLi", "TryHackMe", "OWASP", "web"]
categories: ["Academic Year 2021-22"]

---

# SQL Injection
- Further on today, we will be using THM
- If you haven't registered go to [tryhackme.com](https://tryhackme.com)
- Ensure you have the [VPN file](https://tryhackme.com/access) or AttackBox ready

# SQL
Structured Query Language (SQL) is used to communicate with databases and has the capability to delete, edit, insert or retrieve data. There are a few variations of SQL which have some slight syntactical differences. A few examples are MySQL, SQL Server and SQLite.

{{< admonition type=question open=true >}}
Can you think where databases are used on the web?
{{< /admonition >}}

# SQL Example
What will this do?

```sql
SELECT * FROM Customers WHERE Firstname="Jack";
```

{{< admonition type=tip title="Answer" open=false >}}
It will search the table customers and will return all the people with the first name "Jack".
{{< /admonition >}}

# Injection Attacks

> An injection attack is where a user can trick a server to run a command through an input, header or URL. It is currently in the third position of the [OWASP Top 10](https://owasp.org/Top10/).

Reviewer: There are other ways :)

```suggestion
An injection attack is where a user can trick a server to run a command, via vectors such as an input, header or URL. It is currently in the third position of the [OWASP Top 10](https://owasp.org/Top10/).
```

Author: fair point

{{< /admonition >}}

# A Bad Demo
Can you spot/exploit the flaw?

```py
user=input("Enter Username: ")
password=input("Enter Password: ")
sql = 'SELECT * FROM users WHERE user=="'+user+'"AND password=="'+password+'";'
print("SQL Command Sent: ", sql, "\n")
cursor.execute(sql)
user=cursor.fetchall()
```

{{< admonition type=tip title="How To" open=false >}}
" OR 1=1; --
```py
'SELECT * FROM users WHERE user=="'+user+'"AND password=="'+ password+'";'
```
Becomes:
```sql
SELECT * FROM user WHERE user=="Jack" AND password=="" OR 1=1; --";
```

The command ends the string then compares 1 with 1 (which equals true) and "--" comments out the rest of the statement. In this case, the whole table would be printed.
{{< /admonition >}}

# Can you secure it?
- Research ways that you can secure the program and implement it.

{{< admonition type=tip title="How To" open=false >}}
- Input validation
- Use ? placeholder
- Use the fetchone method - this does not secure it, but makes exploitation much easier
{{< /admonition >}}

# How am I meant to remember the commands?
Thankfully you can use a tool named SQLmap, here is a useful [SQLmap Cheatsheet](https://thedarksource.com/sqlmap-cheat-sheet/).

> It is also possible to pair with Burpsuite, to save having specify location. To do this, intercept the post request, then save it to a file and use the following command:

Reviewer: Could also be a GET request

```suggestion
It is also possible to pair with Burpsuite, to save having specify location. To do this, intercept the request, then save it to a file and use the following command:
```

Author: true

```sh
sqlmap -r filename
```

# If it is that easy...
If it is that easy, why is it so common? Well, many applications are written in PHP which didn't originally contain a SQL validation method. This has meant that many have ignored it. The main problem now is that many insecure code examples exist on code forums/sites.

Another factor is the use of content management systems. This is because if the underlying system is affected, every website using that platform will be vulnerable. However, this can be a benefit, as it will mean that the code only needs to be updated in one place.

> # More Advanced SQLi
> - Time-based/Blind
> - Sometimes you cannot see the output, therefore, you can use timers to show your answer

Reviewer: ```suggestion
# More Advanced SQLi
- Time-based/Blind
    - Sometimes you cannot see the output, therefore, you can use timers to show your answer
- Boolean
    - It is possible to infer the content inside a database using boolean operators, such as changing the order changing the way data is displayed to the user based depending on whether a comparison is true.
```

Author: added, did this manually though as you used "changing" twice. `such as changing the order changing the way `

# NoSQL, no problem?
- Some people believe that no-SQL is injection free, but this is false.
- NoSQL stands for non-relational SQL.
- Although the syntax is different it is still exploitable.
- [Cheat Sheet](https://book.hacktricks.xyz/pentesting-web/nosql-injection).

# Prevent SQLi
- Switch to a static site, like Hugo.

>     - but you lose some functionality (e.g. forms).

Reviewer: You'll know who reads the website based on who messages to point out you have a comment form at the bottom of the page ;)

Author: It's fine, I am more than happy to explain that it is offloaded to a third party. Also, would be nice to know someone actually reads it :)

Reviewer: I read it :))

Reviewer: And I know some people Ive had interviews with have read it :))

Author: Good to know :)
